File: ios/Keeper/Sleeper/SleeperClient.swift
//
//  SleeperClient.swift
//  Keeper
//
//  Created by Han Hossain on 6/30/22.
//

import Foundation
import UIKit

class SleeperClient {
    private let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        return decoder
    }()
    
    func getPlayers() async throws -> [String: Player] {
        let url = URL(string: "https://api.sleeper.app/v1/players/nfl")!
        let (data, _) = try await URLSession.shared.data(from: url)
        
        return try decoder.decode([String: Player].self, from: data)
    }
    
    func getSeasonStatistics() async throws -> [SeasonStatistics] {
        let url = URL(string: "https://api.sleeper.com/stats/nfl/2021?season_type=regular")!
        let (data, _) = try await URLSession.shared.data(from: url)
        
        return try decoder.decode([SeasonStatistics].self, from: data)
    }
    
    func getAvatar(playerId: String) async throws -> UIImage? {
        let uri = playerId.first!.isLetter
            ? "https://sleepercdn.com/images/team_logos/nfl/\(playerId.lowercased()).png"
            : "https://sleepercdn.com/content/nfl/players/\(playerId).jpg"
        
        let url = URL(string: uri)!
        let (data, response) = try await URLSession.shared.data(from: url)
        
        guard
            let httpResponse = response as? HTTPURLResponse,
            httpResponse.statusCode == 200 else {
            return nil
        }
        
        return UIImage(data: data)
    }
}

File: ios-net/Keeper.iOS/Sleeper/SleeperClient.cs
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;
using Keeper.iOS.Sleeper.Models;

namespace Keeper.iOS.Sleeper;

public class SleeperClient
{
    private readonly HttpClient _client;

    private readonly JsonSerializerOptions _options = new JsonSerializerOptions
    {
        PropertyNameCaseInsensitive = true,
        NumberHandling = JsonNumberHandling.AllowReadingFromString
    };

    public SleeperClient(HttpClient httpClient)
    {
        _client = httpClient;
    }

    public async Task<Dictionary<string, SleeperPlayer>> GetPlayersAsync()
    {
        using var response = await _client.GetAsync("https://api.sleeper.app/v1/players/nfl");
        response.EnsureSuccessStatusCode();

        await using var stream = await response.Content.ReadAsStreamAsync();
        return await JsonSerializer.DeserializeAsync<Dictionary<string, SleeperPlayer>>(stream, _options);
    }

    public async Task<List<SleeperSeasonStatistics>> GetSeasonStatisticsAsync()
    {
        using var response = await _client.GetAsync("https://api.sleeper.com/stats/nfl/2021?season_type=regular");
        response.EnsureSuccessStatusCode();

        await using var stream = await response.Content.ReadAsStreamAsync();
        return await JsonSerializer.DeserializeAsync<List<SleeperSeasonStatistics>>(stream, _options);
    }

    public async Task<byte[]> GetAvatarAsync(string playerId)
    {
        var uri = char.IsLetter(playerId[0])
            ? $"https://sleepercdn.com/images/team_logos/nfl/{playerId.ToLower()}.png"
            : $"https://sleepercdn.com/content/nfl/players/{playerId}.jpg";
        using var response = await _client.GetAsync(uri);

        if (response.StatusCode == HttpStatusCode.OK)
        {
            return await response.Content.ReadAsByteArrayAsync();
        }

        return null;
    }
}
